### PR TITLE
Different diff

### PR DESCRIPTION
**Pull request by NightJar**
Resolves issues with the old diff form transformation.

### CODE REVIEW

File: src/Forms/DiffField.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+class DiffField extends HTMLReadonlyField
+{
+    /**
+     * @var FormField Holds the field used as the 'To' (the newer) data for the diff.
+     */
+    protected $comparisonField;
+
+    /**
+     * @param FormField
```
Comment (robbieaverill) [inline src/Forms/DiffField.php]: Missing variable name

File: src/Forms/DiffField.php [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+     * @return $this
+     */
+    public function setComparisonField($field)
+    {
+        if (!$field instanceof FormField) {
+            throw LogicException('$field is not an instance of ' . FormField::class);
```
Comment (robbieaverill) [inline src/Forms/DiffField.php]: Simply typehint FormField instead and let it be a fatal error

File: src/Forms/DiffField.php [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    /**
+     * @return FormField
+     */
+    public function getComparisonField()
+    {
+        return $this->comparisonField;
+    }
+
+    public function Value() {
```
Comment (robbieaverill) [inline src/Forms/DiffField.php]: phpcs

File: src/Forms/DiffField.php [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        $oldValue = $this->getOutdatedField()->Value();
+        $newValue = $this->getComparisonField()->Value();
+        return Diff::compareHTML($oldValue, $newValue);
+    }
+
+    public function getOutdatedField() {
```
Comment (robbieaverill) [inline src/Forms/DiffField.php]: I reckon we could probably use better names. You may not necessarily be comparing an old version against a new version, in fact it could be the opposite to that. In that case, neither field is "outdated." Maybe `getCompareFromField` and `getCompareToField` or even `getFromField` and `getToField`

File: src/Forms/DiffTransformation.php [hunk 5]
```diff
@@ -82,34 +26,23 @@ public function transform(FormField $field)
...
             // First check if a field implements performDiffTransformation()
             parent::transform($field);
         } catch (BadMethodCallException $e) {
-            $diffField = HTMLReadonlyField::create(
-                $name,
-                $field->Title(),
-                Diff::compareHTML($field->Value(), isset($this->data[$name]) ? $this->data[$name] : '')
-            );
-
+            $diffField = DiffField::create($name, $field->Title());
```
Comment (robbieaverill) [inline src/Forms/DiffTransformation.php]: I only just noticed this. We're actually _expecting_ exceptions to be thrown here in order for this class to run its happy path logic? That's wrong... oh well, it is what it is

File: src/Forms/DiffField.php [hunk 6]
```diff
@@ -0,0 +1,76 @@
+<?php
+
+namespace SilverStripe\VersionedAdmin\Forms;
+
+use LogicException;
+use SilverStripe\Forms\FormField;
+use SilverStripe\Forms\HTMLReadonlyField;
+use SilverStripe\View\Parsers\Diff;
+
+class DiffField extends HTMLReadonlyField
```
Comment (robbieaverill) [inline src/Forms/DiffField.php]: Can you please document clearly what this form field is for and provide an inline example of how to use it?

**Comment by robbieaverill** [comment]
@NightJar can you rebase? the first three commits in the list have been merged in another PR. Also can you update your commit messages to reflect what you've changed e.g. "Major refactor of Diff Form Transformation" -> "API Logic for showing diffs between form fields is moved from DiffTransformation to DiffField"

File: src/Forms/DiffField.php (hunk 4 above)
Comment (NightJar) [reply src/Forms/DiffField.php]: Ha, I thought long an hard about this. I had from and to and the class got a bit unmanageable (keeping track of two fields instead of one). In the end I considered @ clarkepaul's request that comparisons always be older to newer (chronologically), but you are correct in that this form field could be used elsewhere.

>`diff [-c| -e| -f| -C n][-br] file1 file2`
>
>**DESCRIPTION**
>The diff utility shall compare the contents of `file1` and `file2` and write to standard output a list of changes necessary _to convert `file1` into `file2`_. This list should be minimal. No output shall be produced if the files are identical.

I settled on 'outdated' in the manner that a `diff(1)` is always what it takes to turn `from` _into_ `to`, leading me to think in terms of `outdated` to `updated`, even if that update is a reversal... kinda like accounting (or say, `git`, you undo by re-doing in the negative, as such).

**tl;dr**
So in that regard `outdated` is 'old' not in terms of chronology, but in terms of _succession_.

Would you be happy to settle for a comment (/docblock) summarising this?

Comment (robbieaverill) [reply src/Forms/DiffField.php]: >In the end I considered @ clarkepaul's request that comparisons always be older to newer (chronologically), but you are correct in that this form field could be used elsewhere.

Yeah, that's a UI request AFAIK. This field will probably end up in core and could be used for other things (like advanced workflow)

Comment (robbieaverill) [reply src/Forms/DiffField.php]: Yeah, doc blocks would be good

File: src/Forms/DiffTransformation.php (hunk 5 above)
Comment (NightJar) [reply src/Forms/DiffTransformation.php]: Yeah, the nature of `FormTransformation` is that they actually call `$field->transform($transformation)` rather than having _anything_ to do with `Form`.

In turn `$field->transform($transformation)` calls `$transformation->transform($this)` if and only if there is not a function intended for use with higher precedence (`$field->performDiffTransformation()`)
...so the transformation (class) has nothing at all to do with the `FormField`, either.

There are no hook points, so we rely on the field being trasnformed not knowing what to do with a DiffTransformation first, in order to not duplicate the logic between `FormField` and this `DiffTransformation`.

The whole thing is a bit messy, but indeed, it is what it is.

**Comment by NightJar** [comment]
Done
